File: 61. Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
		if (head == nullptr || k == 0) return head;

		ListNode *p = head, *tail = nullptr;
		int len = 0;
		while (p) {
			len++;
			tail = p;
			p = p->next;
		}
		k %= len;
		int cnt = len - k;
		p = head;
		while (cnt > 1) {
			p = p->next;
			cnt--;
		}

		tail->next = head;
		if (p) {
			head = p->next;
			p->next = nullptr;
		}
			
		return head;
    }
};

File: go/src/leetcode/61. Rotate List/61. Rotate List.go
package main

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil {
		return head
	}
	n := 0
	for p := head; p != nil; p = p.Next {
		n++
	}
	k %= n
	if k == 0 {
		return head
	}
	dummy := new(ListNode)
	dummy.Next = head
	p := dummy
	for i := 0; i != n-k; i++ {
		p = p.Next
	}
	head = p.Next
	p.Next = nil
	p = head
	for p.Next != nil {
		p = p.Next
	}
	p.Next = dummy.Next
	return head
}

type ListNode struct {
	Val int
	Next *ListNode
}
